codecov_auth/tests/unit/views/test_okta_cloud.py

```diff
@@ -0,0 +1,453 @@
+from typing import Any
+from urllib.parse import unquote, urlencode, urlparse
+
+import pytest
+from django.test import override_settings
+from pytest import LogCaptureFixture
+from pytest_mock import MockerFixture
+from shared.django_apps.codecov_auth.models import Account, OktaSettings, Owner
+from shared.django_apps.codecov_auth.tests.factories import (
+    AccountFactory,
+    OktaSettingsFactory,
+)
+
+from codecov_auth.tests.factories import OwnerFactory
+from codecov_auth.views.okta_cloud import (
+    OKTA_CURRENT_SESSION,
+    OKTA_SIGNED_IN_ACCOUNTS_SESSION_KEY,
+)
+from codecov_auth.views.okta_mixin import OktaIdTokenPayload
+from utils.test_utils import Client as TestClient
+
+
+@pytest.fixture
+def signed_in_client() -> TestClient:
+    new_client = TestClient()
+    new_client.force_login_owner(OwnerFactory())
+    return new_client
+
+
+@pytest.fixture
+def okta_org_name() -> str:
+    return "foo-bar-organization"
+
+
+@pytest.fixture
+def okta_org(okta_org_name: str) -> Owner:
+    org: Owner = OwnerFactory.create(username=okta_org_name, service="github")
+    org.save()
+    return org
+
+
+@pytest.fixture
+def okta_account(okta_org: Owner):
+    account = AccountFactory()
+    okta_org.account = account
+    okta_org.save()
+
+    okta_settings: OktaSettings = OktaSettingsFactory(account=account)
+    okta_settings.url = "https://foo-bar.okta.com"
+    okta_settings.save()
+    return account
+
+
+@pytest.fixture
+def mocked_okta_token_request(mocker):
+    return mocker.patch(
+        "codecov_auth.views.okta_mixin.requests.post",
+        return_value=mocker.MagicMock(
+            status_code=200,
+            json=mocker.MagicMock(
+                return_value={
+                    "access_token": "test-access-token",
+                    "refresh_token": "test-refresh-token",
+                    "id_token": "test-id-token",
+                    "state": "test-state",
+                },
+            ),
+        ),
+    )
+
+
+@pytest.fixture
+def mocked_validate_id_token(mocker):
+    return mocker.patch(
+        "codecov_auth.views.okta_cloud.validate_id_token",
+        return_value=OktaIdTokenPayload(
+            sub="test-id",
+            email="test@example.com",
+            name="Some User",
+            iss="https://example.com",
+            aud="test-client-id",
+        ),
+    )
+
+
+def log_message_exists(message: str, logs: list[LogRecord]) -> bool:
+    """Helper method to check that a particular log record was emitted"""
+    for log in logs:
+        if log.message == message:
+            return True
+    return False
+
+
+@pytest.mark.django_db
+def test_okta_login_unauthenticated_user(
+    client: TestClient,
+    caplog: LogCaptureFixture,
+):
+    res = client.get("/login/okta/github/some-unknown-service")
+    assert log_message_exists(
+        "User needs to be signed in before authenticating organization with Okta.",
+        caplog.records,
+    )
+    assert res.status_code == 403
```

Comment:
does it make sense to also (or instead) include those messages as part of the response, instead of just making those generic 403 pages?

Reply:
Hmm, it's a tradeoff between giving users too much info for attackers to learn about our system and usability for users. For authentication APIs, I would lean more conservatively towards exposing less information automatically to users (and adjust based on user feedback). For example, the log message above exposes that Codecov has a 2 step authentication process involving Okta, and that this endpoint is supposed to authenticate users for specific organizations (rather than a user generically).